manual/manual/html_processing/src/process_manual.ml

```diff
@@ -0,0 +1,481 @@
+  let new_html = to_string soup in
+  write_file (docs_file file) new_html
+
+(* Find title associated with file *)
+let file_title file toc =
+  if file = "index.html" then Some index_title
+  else Hashtbl.find_opt toc file
+
+(* Replace the images of one of the "previous, next, up" link by the title of
+   the reference. *)
+let nav_replace_img_by_text toc alt a img =
+  let file = R.attribute "href" a in
+  let title = match file_title file toc with
+    | Some f -> begin match alt with
+        | "Previous" -> "« " ^ f
+        | "Next" -> f ^ " »"
+        | "Up" -> f
+        | _ -> failwith "This should not happen"
+            end
+    | None -> dbg "Unknown title for file %s" file; file in
+  let txt = create_text title in
+  replace img txt;
+  add_class (String.lowercase_ascii alt) a
+
+(* Replace three links "Previous, Up, Next" at the end of the file by more
+   useful titles, and insert then in a div container, keeping only 2 of them:
+   either (previous, next) or (previous, up) or (up, next). Remove them at the
+   top of the file, where they are not needed because we have the TOC. *)
+let update_navigation soup toc =
+  Option.iter delete (soup $? "hr");
+  let links =
+    ["Previous"; "Up"; "Next"]
+    |> List.map (fun alt -> alt, to_list (soup $$ ("img[alt=\"" ^ alt ^ "\"]")))
+    (* In principle [imgs] will contain either 0 or 2 elements. *)
+    |> List.filter (fun (_alt, imgs) -> List.length imgs = 2)
+    (* We delete the first link, and replace image by text *)
+    |> List.map (fun (alt, imgs) ->
+        delete (R.parent (List.hd imgs));
+        let img = List.hd (List.rev imgs) in
+        let a = R.parent img in
+        nav_replace_img_by_text toc alt a img;
+        a) in
+  if links <> [] then begin
+    (* We keep only 2 links: first and last *)
+    let a1, a2 = match links with
+      | [prev;up;next] -> delete up; (prev, next)
+      | [a;b] -> (a,b)
+      | _ -> failwith "Navigation link should have at least 2 elements" in
+    add_class "previous" a1;
+    add_class "next" a2;
+    (* some elements can have both previous and up classes, for instance. This
+       helps css styling. *)
+    let container = create_element ~class_:"bottom-navigation" "div" in
+    wrap a1 container;
+    append_child container a2
+  end
+
+
+(* extract the cut point (just after title) and the header of soup:
+   "clone_structure" needs them to insert external files after the cut point,
+   and include the TOC. *)
+let make_template soup =
+  let header = soup $ "header" in
+  let title = match soup $? "div.maintitle" with
+    | Some div -> div (* This is the case for "index.html" *)
+    | None -> soup $ "h1" in
+  title, header
+  
+(* Create a new file by keeping only the head/headers parts of "soup", deleting
+   everything after the title, and inserting the content of external file (hence
+   preserving TOC and headers) (WARNING: this mutates soup) *)
+let clone_structure soup (title, header) toc xfile =
```

Comment:
The name `clone_structure` does not seems to fit the notion of the function. Maybe `insert_xfile_content`?

Reply:
done